import "common.proto";
import "type.proto";
import "tablet.proto";
package rtidb.ObjectStorage;

option cc_generic_services = true;
option java_generic_services = true;
option java_package = "com._4paradigm.rtidb.object_storage_server";
option java_outer_classname = "ObjectStorage";

message GeneralResponse {
    optional string msg = 1;
    optional int32 code = 2;
};

message GeneralRequest {
    optional uint32 tid = 1;
    optional uint32 pid = 2;
    optional string key = 3;
}

message GetResponse {
    optional bytes pairs = 1;
    optional string msg = 2;
    optional int32 code = 3;
}

message PutRequest {
    optional uint32 tid = 1;
    optional uint32 pid = 2;
    optional bytes pairs = 3;
}

message OssTableMeta {
    optional uint32 tid = 1;
    optional string name = 2;
    optional uint32 pid = 3;
    optional rtidb.common.StorageMode storage_mode = 4 [default = kHDD];
    optional rtidb.type.TableType table_type = 5;
    optional rtidb.api.CompressType compress_type = 6;
    optional uint32 height = 7 [default = 1];
    optional int64 before_time = 8 [default = 0];
}

message CreateTableRequest {
    optional OssTableMeta table_meta = 1;
}

message StatsRequest {
    optional uint32 tid = 1;
    optional uint32 pid = 2;
}

message StatsResponse {
    optional string msg = 1;
    optional int32 code = 2;
    optional uint64 count = 3;
    optional uint64 total_space = 4;
    optional uint64 avail_space = 5;
}

service OSS {
    rpc CreateTable(CreateTableRequest) returns (GeneralResponse);
    rpc Put(PutRequest) returns (GeneralResponse);
    rpc Get(GeneralRequest) returns (GetResponse);
    rpc Delete(GeneralRequest) returns (GeneralResponse);
    rpc Stats(StatsRequest) returns (StatsResponse);
}